### PR TITLE
Add support for dynamic rule detection for pipeline config transformation

### LINKED ISSUE #4600: Dynamic Rule Detection
**Is your feature request related to a problem? Please describe.**
Dynamic Rule Detection for pipeline configuration transformation. 

**Describe the solution you'd like**
Scan rule folder. Apply each rule. Check whether the rule is valid or not, if it is valid, it should return a corresponding template.

**Additional context**
Extension to issue #4446

### PR DESCRIPTION
**Pull request by srikanthjg**
### Description
Dynamically detects rules present in rules folder and applies a corresponding transformation based on plugin name.
 
### Issues Resolved
Resolves #4600 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by kkondaka — APPROVED** [review]

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java [hunk 1]
```diff
@@ -41,35 +53,63 @@ public String getTransformationRulesDirectoryLocation() {
...
+        URI uri = null;
+        try {
+            uri = getClass().getClassLoader().getResource("rules").toURI();
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        Path rulesFolderPath;
+
+        if ("jar".equals(uri.getScheme())) {
```
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:82]: in runtime, we would expect everything inside a jar.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java [hunk 2]
```diff
@@ -41,35 +53,63 @@ public String getTransformationRulesDirectoryLocation() {
...
     }
 
-    public InputStream getPluginRuleFileStream(String pluginName) {
-        if(pluginName == null || pluginName.isEmpty()){
-            throw  new RuntimeException("Transformation plugin not found");
+    public List<Path> getRuleFiles() {
+        // Get the URI of the rules folder
+        URI uri = null;
+        try {
+            uri = getClass().getClassLoader().getResource("rules").toURI();
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:75]: Let's use a predefined directory structure to ensure that we are only getting the transform rules:

```
org/opensearch/dataprepper/transforms/
```

Thus, we'd have:

```
org/opensearch/dataprepper/transforms/rules/
org/opensearch/dataprepper/transforms/templates/
```

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java [hunk 3]
```diff
@@ -41,35 +53,63 @@ public String getTransformationRulesDirectoryLocation() {
...
+            // File is inside a JAR, create a filesystem for it
+            try (FileSystem fileSystem = FileSystems.newFileSystem(uri, Collections.emptyMap())) {
+                rulesFolderPath = fileSystem.getPath("rules");
+                return scanFolder(rulesFolderPath);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        } else {
+            // File is not inside a JAR
+            rulesFolderPath = Paths.get(uri);
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:92]: When would this ever occur?

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java [hunk 4]
```diff
@@ -41,35 +53,63 @@ public String getTransformationRulesDirectoryLocation() {
...
-        return rulesDirectoryPath + "/" + pluginName + RULE_FILE_NAME_PATTERN;
+        ClassLoader classLoader = TransformersFactory.class.getClassLoader();
+        InputStream filestream = classLoader.getResourceAsStream("templates" + "/" + pluginName + TEMPLATE_FILE_NAME_PATTERN);
+        return filestream;
     }
 
-    public InputStream getPluginRuleFileStream(String pluginName) {
-        if(pluginName == null || pluginName.isEmpty()){
-            throw  new RuntimeException("Transformation plugin not found");
+    public List<Path> getRuleFiles() {
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:71]: This API contract lists out `Path`s for rules. Then the caller comes back and get the input stream by calling `readRuleFile`. You should be able to consolidate these. 

Just make one method: `Collection<InputStream> loadRules()`.

I do see that you use the file name later, I think for debugging purposes. You could make a class that holds both a name and an InputStream.

```
class RuleInputStream {
  private String name;
  private InputStream inputStream
}
```

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleFileEvaluation.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+@Builder(setterPrefix = "with")
+@AllArgsConstructor
+@Data
+public class RuleFileEvaluation {
+    Boolean result;
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleFileEvaluation.java:11]: These should all be private. And I think they can also be private.

File: data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml [hunk 6]
```diff
@@ -1,3 +1,4 @@
+plugin_name: "documentdb"
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml:1]: As this is for testing, let's add test in the name. Maybe `test-documentdb`.

File: data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb1-rule.yaml [hunk 7]
```diff
@@ -1,2 +1,3 @@
+plugin_name: "documentdb"
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb1-rule.yaml:1]: Similar comment as above, please add `test` into the name.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java [hunk 8]
```diff
@@ -94,38 +81,63 @@ private RuleEvaluatorResult isDocDBSource(PipelinesDataFlowModel pipelinesModel)
...
+
+                for (String rule : rules) {
+                    try {
+                        JsonNode result = JsonPath.using(parseConfig).parse(pipelinesJson).read(rule);
+                        if (result == null || result.size()==0) {
+                            allRulesValid = false;
+                            break;
+                        }
+                    } catch (PathNotFoundException e) {
+                        LOG.debug("Json Path not found for {}", ruleFile.getFileName().toString());
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:116]: No need to call `.toString()` here.

File: data-prepper-pipeline-parser/src/test/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluatorTest.java [hunk 9]
```diff
@@ -31,16 +34,16 @@
...
 
-        // Set up
         String ruleDocDBFilePath = TestConfigurationProvider.RULES_TRANSFORMATION_DOCUMENTDB_CONFIG_FILE;
+        String ruleDocDBTemplatePath = TestConfigurationProvider.TEMPLATE_TRANSFORMATION_DOCDB1_CONFIG_FILE;
         String pluginName = "documentdb";
         String pipelineName = "test-pipeline";
-        Map sourceOptions = new HashMap<String, Object>();
-        Map s3_bucket = new HashMap<>();
+        Map<String, Object> sourceOptions = new HashMap<>();
+        Map<String, Object> s3_bucket = new HashMap<>();
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/test/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluatorTest.java:44]: Please use correct naming conventions: `s3Bucket`.

File: data-prepper-pipeline-parser/src/test/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluatorTest.java [hunk 10]
```diff
@@ -51,54 +54,60 @@ void test_isTransformationNeeded_ForDocDBSource_ShouldReturn_True() throws FileN
...
+    void test_isTransformationNeeded_ForOtherSource_ShouldReturn_False() throws IOException {
+
+        String pluginName = "http";
         String pipelineName = "test-pipeline";
-        Map sourceOptions = new HashMap<String, Object>();
-        sourceOptions.put("option1", "1");
-        sourceOptions.put("option2", null);
-        final PluginModel source = new PluginModel("http", sourceOptions);
+        Map<String, Object> sourceOptions = new HashMap<>();
+        Map<String, Object> s3_bucket = new HashMap<>();
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/test/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluatorTest.java:82]: Similar comment about variable naming conventions.

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Comment by kkondaka** [comment]
@srikanthjg Please address David's other comments. I will approve again and we can ask Hai to review too.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java (hunk 2 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:75]: The way i understand this is, we want to move the template and rules to be plugin specific and move this into data-prepper-plugin directory, ie, each plugin gets a rule and template but there are some limitations to this approach. If for example, there is an intent by an user for which plugin does not exist and we want to tranform that to other processors, where do we place the rules and templates in this case. Example for this is with respect to security lake.

**Comment by kkondaka** [comment]
@srikanthjg I am not sure if we are going to have a "real" plugin for each of the plugins in the pre-transformation config file. So, while moving documentdb related rules and templates to document db directory is a good suggestion by David, it might not be possible in all cases. For now, I suggest we just commit what you implemented in this PR and if needed, we can take the approach suggested by David in a future PR

File: data-prepper-pipeline-parser/src/main/resources/rules/documentdb-rule.yaml [hunk 11]
```diff
@@ -1,3 +1,4 @@
+plugin_name: "documentdb"
```
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/main/resources/rules/documentdb-rule.yaml:1]: https://github.com/opensearch-project/data-prepper/pull/4601#discussion_r1704527874

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java (hunk 3 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:92]: I was trying to address 2 cases:
1. "jar" scheme: This case occurs when the code is running from within a JAR file, ie during runtime.
2. Non-"jar" scheme: This case occurs when the code is running in a development environment where the rules directory is available in the file system and not within a jar.

File: data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml (hunk 6 above)
Comment (kkondaka) [reply data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml:1]: @srikanthjg do you want to make this small change?

Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml:1]: I think it will be inconsistent even within the same rule yaml. The apply_when condition for the rule yaml, we still define it as documentdb and it would looks as follows: 
```
plugin_name: "documentdbtest"
apply_when:
  - "$..source.documentdb"
 ```
We still want use documentdb and not documentdbtest in the apply_when condition as that is what is used in the pipeline yaml; and that cannot be mocked that easily.

The plugin name in the rule yaml is only used to get the corresponding template, but it seems inconsistent if we use documentdb vs documentdbtest even within the same file for the same plugin we want to test.

File: data-prepper-plugins/dummy-plugin/build.gradle
Comment (dlvenable) [inline data-prepper-plugins/dummy-plugin/build.gradle:1]: You can remove this empty file. You don't need a `build.gradle` file at all.

File: data-prepper-plugins/dummy-plugin/src/main/resources/org/opensearch/dataprepper/transforms/rules/dummy-plugin-rule.yaml
Comment (dlvenable) [inline data-prepper-plugins/dummy-plugin/src/main/resources/org/opensearch/dataprepper/transforms/rules/dummy-plugin-rule.yaml:1]: Please add a `README.md` to this project to explain how to use it and test with it.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java [hunk 12]
```diff
@@ -34,52 +34,39 @@ public class RuleEvaluator {
...
                     return RuleEvaluatorResult.builder()
                             .withEvaluatedResult(true)
                             .withPipelineTemplateModel(templateModel)
                             .withPipelineName(entry.getKey())
                             .build();
                 }
-            } catch (FileNotFoundException e){
-                LOG.error("Template File Not Found for {}", PLUGIN_NAME);
+            } catch (FileNotFoundException e) {
+                LOG.error("Template File Not Found");
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:67]: Why did you remove the additional information for tracking the problem?

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java [hunk 13]
```diff
@@ -94,47 +81,68 @@ private RuleEvaluatorResult isDocDBSource(PipelinesDataFlowModel pipelinesModel)
...
+                    }
+                } finally {
                     ruleStream.close();
-                } catch (IOException e) {
-                    throw new RuntimeException(e);
                 }
             }
+
+        } catch (FileNotFoundException e) {
+            LOG.debug("Rule File Not Found");
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:131]: Should this be an error?

Also, can you include the file name in the log?

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java [hunk 14]
```diff
@@ -4,84 +4,116 @@
...
+                URL rulesURL = rulesURLs.nextElement();
+
+                try {
+                    // Convert the URL to a URI, then to a Path to read the directory contents
+                    Path rulesPath;
+                    try {
+                        rulesPath = Paths.get(rulesURL.toURI());
+                    } catch (FileSystemNotFoundException e) {
+                        // Handle the case where the file system is not accessible (e.g., in a JAR)
+                        FileSystem fileSystem = FileSystems.newFileSystem(rulesURL.toURI(), Collections.emptyMap());
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:92]: Is this going to work? These will always be in jar files.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml (hunk 6 above)
Comment (dlvenable) [reply data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml:1]: Why wouldn't this work?

```
plugin_name: "documentdbtest"
apply_when:
  - "$..source.documentdbtest"
```

These files are only used for unit testing, right? So, there is no need to keep the names matching the others.

Comment (dlvenable) [reply data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml:1]: This is where you use it:

https://github.com/opensearch-project/data-prepper/pull/4601/files#diff-40ef318469b1b8fb28c6ee16a998f746921ecee5471a9f481d497617b60a2180R38

File: data-prepper-plugins/dummy-plugin/build.gradle
Comment (srikanthjg) [reply data-prepper-plugins/dummy-plugin/build.gradle:1]: build.gradle is needed to recognize the module. When i tried without build.gradle, classloader search was not picking up the rule file from the folder.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java (hunk 14 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/TransformersFactory.java:92]: this works. It will always hit this code when dataprepper is running because it is always in jar during execution.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java (hunk 12 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:67]: will add it back.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java [hunk 15]
```diff
@@ -64,7 +65,7 @@ public RuleEvaluatorResult isTransformationNeeded(PipelinesDataFlowModel pipelin
                             .build();
                 }
             } catch (FileNotFoundException e) {
-                LOG.error("Template File Not Found");
+                LOG.error("Template File Not Found for {}", PLUGIN_NAME);
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:68]: Actually, you don't even need this catch block. You can remove it. And then go back to using a local `pluginName` variable.

File: data-prepper-plugins/dummy-plugin/README.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+template to transform to when the rule is valid.
+
+For further details on transformation refer:
+/docs/pipeline_configuration_transformation.md
+
+
+For Example:
+
+User Config:
+```aidl
```
Comment (dlvenable) [inline data-prepper-plugins/dummy-plugin/README.md:15]:
```suggestion
```yaml
```

File: data-prepper-plugins/dummy-plugin/README.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  sink:
+    - stdout:
+```
+
+Here dummy_plugin is not really a plugin that is defined in dataprepper, but we can 
+use the pipeline transformation to convert the user-config to different config based 
+on the template.
+
+Rule:
+```aidl
```
Comment (dlvenable) [inline data-prepper-plugins/dummy-plugin/README.md:33]:
```suggestion
```yaml
```

File: data-prepper-plugins/dummy-plugin/README.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+plugin_name: "dummy-plugin"
+apply_when:
+  - "$.test-pipeline.processor[?(@.dummy_plugin)]"
+```
+
+Template:
+```aidl
```
Comment (dlvenable) [inline data-prepper-plugins/dummy-plugin/README.md:40]:
```suggestion
```yaml
```

File: data-prepper-plugins/dummy-plugin/README.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+      path: "/tmp/input-file.log"
+  processor:
+    - string_converter:
+        upper_case: true
+  sink:
+    - noop:
+```
+
+Output:
+```aidl
```
Comment (dlvenable) [inline data-prepper-plugins/dummy-plugin/README.md:53]:
```suggestion
```yaml
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml (hunk 6 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/test/resources/transformation/rules/documentdb-rule.yaml:1]: i will need to change like 20 files(test yaml files) in the test as all of them use documentdb. i cant just change the rule to documentdbtest and make it work. The userconfig, expected output and template depends on this. can i add this as another PR? it should it mostly replace all but i see some other issues with it.

**Review by dlvenable — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]
